certstore/certstore.go

```diff
@@ -121,20 +174,104 @@ func (cs *Store) GetRange(ctx context.Context, start uint64, end uint64) ([]cert
+	}
+	return nil
+}
+
+func (cs *Store) GetPowerTable(ctx context.Context, instance uint64) (gpbft.PowerEntries, error) {
+	if instance < cs.firstInstance {
+		return nil, fmt.Errorf("cannot return a power table before the first instance: %d", cs.firstInstance)
+	}
+	if instance > cs.Latest().GPBFTInstance+1 {
+		return nil, fmt.Errorf("cannot return future power table for instance %d", instance)
+	}
+
+	// We store every `powerTableFrequency` power tables. Find the nearest multiple smaller than
+	// the requested instance.
+	startInstance := instance
+	if offset := instance % cs.powerTableFrequency; offset != 0 {
+		startInstance -= offset
+	}
+	// Walk backwards trying to find a starting power table.
+	var powerTable gpbft.PowerEntries
+	for startInstance >= cs.firstInstance {
+		pt, err := cs.ds.Get(ctx, cs.keyForPowerTable(instance))
+		if err == nil {
+			err = powerTable.UnmarshalCBOR(bytes.NewReader(pt))
+		} else if errors.Is(err, datastore.ErrNotFound) {
+			startInstance -= cs.powerTableFrequency
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(powerTable) == 0 {
+		return nil, xerrors.Errorf("no power tables found")
+	}
+	// Load the power table diffs up till (but not including) the target instance.
+	certificates, err := cs.GetRange(ctx, startInstance, instance-1)
+	if err != nil {
+		return nil, err
+	}
+
+	// Apply the diffs, saving all power tables that are a multiple of `powerTableFrequency` so
+	// we don't have to walk back so far next time.
+	deltas := make([][]certs.PowerTableDelta, 0, len(certificates))
+	for i := range certificates {
+		cert := &certificates[i]
+		deltas = append(deltas, cert.PowerTableDelta)
+		if cert.GPBFTInstance%cs.powerTableFrequency == cs.powerTableFrequency-1 {
+			powerTable, err = certs.ApplyPowerTableDiffs(powerTable, deltas...)
+			if err != nil {
+				return nil, xerrors.Errorf("applying power deltas: %w", err)
+			}
+
+			if err := cs.putPowerTable(ctx, cert.GPBFTInstance, powerTable); err != nil {
+				return nil, err
+			}
+			deltas = deltas[:0]
+		}
+	}
+	powerTable, err = certs.ApplyPowerTableDiffs(powerTable, deltas...)
+	if err != nil {
+		return nil, xerrors.Errorf("applying power deltas: %w", err)
+	}
+	return powerTable, err
+}
+
+func (*Store) keyForCert(i uint64) datastore.Key {
 	return datastore.NewKey(fmt.Sprintf("/certs/%016X", i))
 }
 
+func (*Store) keyForPowerTable(i uint64) datastore.Key {
+	return datastore.NewKey(fmt.Sprintf("/power/%016X", i))
+}
+
 // Put saves a certificate in a store and notifies listeners.
-// It errors if adding a cert would create a gap
+// It errors if adding a cert would create a gap.
 func (cs *Store) Put(ctx context.Context, cert *certs.FinalityCertificate) error {
-	key := cs.keyForInstance(cert.GPBFTInstance)
+	if cert.GPBFTInstance < cs.firstInstance {
+		return xerrors.Errorf("certificate store only stores certificates on or after instance %d", cs.firstInstance)
+	}
+	cs.writeLk.Lock()
+	defer cs.writeLk.Unlock()
```

Comment:
I've moved the lock up to the top because we really don't care about contention on write (there should only ever be one writer).